orderer/common/cluster/util.go

```diff
@@ -793,3 +797,33 @@ func (cm *ComparisonMemoizer) setup() {
 	cm.rand = rand.New(rand.NewSource(time.Now().UnixNano()))
 	cm.cache = make(map[arguments]bool)
 }
+
+func exportKM(cs tls.ConnectionState, label string, context []byte) ([]byte, error) {
+	tlsBinding, err := cs.ExportKeyingMaterial(label, context, 32)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed generating TLS Binding material")
+	}
+	return tlsBinding, nil
+}
+
+func GetTLSSessionBinding(stream grpc.Stream, signPayload *orderer.AuthRequest) ([]byte, error) {
+	peerInfo, ok := peer.FromContext(stream.Context())
+	connState := peerInfo.AuthInfo.(credentials.TLSInfo).State
+	if !ok {
+		return nil, errors.New("failed extracting stream context")
+	}
+
+	encodeSignFields, err := protoutil.Marshal(signPayload)
```

Comment:
Protobuf is not a deterministic library and the RFC clearly states to use: `r = Hash(version || timestamp || from_id || to_id, channel)` so let's just compute the hash outside and pass it instead of passing the entire AuthRequest.

Reply:
do you mean concat the fields individually by referring to this operator `||`?, like below:

```
	bindingFieldsHash := util.ComputeSHA256(util.ConcatenateBytes(
		[]byte(strconv.FormatUint(uint64(version), 10)),
		[]byte(timestamp.String()),
		[]byte(strconv.FormatUint(fromId, 10)),
		[]byte(strconv.FormatUint(toId, 10)),
		[]byte(stream.Channel),
	))
```

Reply:
yes